File: Nim/Kex_1sub.nim
import os
import random
import times
from strutils import parseInt
import Player
import Golem

var GOLEM_HIT = 70
var GOLEM_FLEE = 20
var ATTACK_HIT = 110
var SKILL_HIT = 100
var MAGIC_HIT = 70

proc check_hit_or_miss(hit: int): bool =
    randomize(getTime().toWinTime)
    var val = rand(99)
    return val <= hit - 1

proc attack_golem_ex*(lv: int, player: var Player) =
    var golem = GolemEx(
        hp: lv * 50 + 100,
        defense: lv * 10 + 40,
        attack: lv * 10 + 40,
        hit: GOLEM_HIT,
        flee: GOLEM_FLEE
    )

    echo "ゴーレムLv.", lv + 1, "が現れた！"
    while golem.hp > 0:
        echo "ゴーレム 残りHP：", golem.hp
        stdout.write "武器を選択してください（１．攻撃　２．特技　３．魔法）＞"
        stdout.flushFile
        var choice = parseInt(readLine(stdin))
        randomize(getTime().toWinTime)
        var damage: int
        var is_hit = false

        case choice
        of 1:
            damage = 60 + rand(40)
            is_hit = check_hit_or_miss(ATTACK_HIT - golem.flee)
        of 2:
            damage = 30 + rand(100)
            is_hit = check_hit_or_miss(SKILL_HIT - golem.flee)
        of 3:
            damage = 20 + rand(180)
            is_hit = check_hit_or_miss(MAGIC_HIT - golem.flee)
        else:
            damage = 0
        
        if is_hit:
            damage -= golem.defense
            if damage <= 0:
                damage = 0
            echo damage, "のダメージ！"
            golem.hp -= damage
        else:
            echo "攻撃を外した！"
        
        echo "ゴーレムの攻撃！"
        is_hit = check_hit_or_miss(golem.hit)
        if is_hit:
            var injury = golem.attack - player.defense
            echo injury, "のダメージ！"
            player.hp -= injury
            if player.hp <= 0:
                player.hp = 0
                echo "あなたはゴーレムに負けました！ゲームオーバー！"
                return
        else:
            echo "攻撃を外した！"
        
        echo "プレイヤー残りHP：", player.hp
        if golem.hp <= 0:
            golem.hp = 0
        if golem.hp == 0:
            echo "ゴーレムLv.", lv + 1, "を倒した！"

File: Dart/lib/Kex_1sub.dart
import 'dart:io';
import 'dart:math';
import 'package:Dart/Golem.dart';
import 'package:Dart/Player.dart';

const int ATTACK_HIT = 110;
const int SKILL_HIT = 100;
const int MAGIC_HIT = 70;

bool CheckHitOrMiss(int hit) {
  Random rand = Random();
  int val = rand.nextInt(100);
  return val <= hit - 1;
}

void AttackGolemEx(int lv, Player player) {

  GolemEx golem = GolemEx(lv * 50 + 100, lv * 10 + 40, lv * 10 + 40);
  print('ゴーレムLv.${lv + 1}が現れた！');
  while (golem.Hp > 0) {
    print('ゴーレム 残りHP：${golem.Hp}');
    stdout.write('武器を選択してください（１．攻撃　２．特技　３．魔法）＞');
    Random rand = Random();
    int damage = 0;
    bool is_hit = false;

    switch (int.parse(stdin.readLineSync())) {
      case 1:
        damage = 60 + rand.nextInt(41);
        is_hit = CheckHitOrMiss(ATTACK_HIT - golem.Flee);
        break;
      case 2:
        damage = 30 + rand.nextInt(101);
        is_hit = CheckHitOrMiss(SKILL_HIT - golem.Flee);
        break;
      case 3:
        damage = 20 + rand.nextInt(181);
        is_hit = CheckHitOrMiss(MAGIC_HIT - golem.Flee);
        break;
      default:
        break;
    }

    if (is_hit) {
      damage -= golem.Defense;
      if (damage <= 0) {
        damage = 0;
      }
      print('$damageのダメージ！');
      golem.Hp -= damage;
    } else {
      print('攻撃を外した！');
    }

    print('ゴーレムの攻撃！');
    is_hit = CheckHitOrMiss(golem.Hit);
    if (is_hit) {
      int injury = golem.Attack - player.Defense;
      print('$injuryのダメージ！');
      player.Hp -= injury;
      if (player.Hp <= 0) {
        player.Hp = 0;
        print('あなたはゴーレムに負けました！ゲームオーバー！');
        return;
      }
    } else {
      print('攻撃を外した！');
    }

    print('プレイヤー残りHP：${player.Hp}');
    if (golem.Hp <= 0) {
      golem.Hp = 0;
    }
    if (golem.Hp == 0) {
      print('ゴーレムLv.${lv + 1}を倒した！');
    }
  }
  return;
}